### PR TITLE
Deduplicate strings in binlogs

### PR DESCRIPTION
**Pull request by KirillOsenkov**
When writing out a binary log we now deduplicate strings and dictionaries. This results in a significant performance increase and binlog size reduction. Performance increase is about 2x on average, size reduction is about 4x on average, but up to 30x for large binlogs.

Add two new record kinds: String and NameValueList. A string record is written the first time we encounter a string we need to serialize. The next time we see the string we only write its index in the total list of strings.

Similarly, NameValueList is a list of key and value strings, used for Properties, environment variables and Item metadata. The first time we're writing out a list we write a record, and subsequent times just the index.

This keeps the binlog format streaming, so if the logging is interrupted midway, we will be able to read everything up to that point.

We do not hold on to strings we encountered. Instead we hash them and only preserve the hash code. We rely on the hash function resulting in no collisions, otherwise it could happen that a random string in a binlog would be substituted instead of another one. The hashtables do not significantly add to MSBuild memory usage (20-30 MB for our largest binlogs).

FNV-1a (64-bit hash size) was a carefully chosen hash function for its simplicity, performance, and lack of collisions on all binlogs tested so far. 32-bit hash size (such as string.GetHashCode() was not sufficient and resulted in ~800 collisions for our largest binlog with 2.7 million strings.

This change includes other performance improvements such as inserting a BufferedStream around the stream we're reading or writing. This results in a significant performance improvement.

We introduce a new StringStorage data structure in the binlog reader, for storing the strings on disk instead of reading them all into memory. Strings are loaded into memory on demand. This prevents OOM in 32-bit MSBuild processes when playing back large binlogs. This keeps the memory usage relatively flat when reading.

### Testing

I've done extensive testing for correctness, performance, and hash collisions and performance.

Hash function selection is documented here:
https://github.com/KirillOsenkov/MSBuildStructuredLog/wiki/String-Hashing
In summary, Fnv-1a provides no collisions with 64-bit hash size, and very decent performance (11 seconds to hash all the strings in our largest binlog, 2.7 million strings). Additionally it's very simple and minimal code.

Playback:
I've tested various scenarios for playback:
 1. binlog format
   - playing back a legacy binlog (format version < 10), with no String or NameValueList records
   - playing back version 10 binlog
 2. process bitness
   - on 32-bit we need to use the StringStorage helper class when reading new binlogs, to avoid OOM
   - on 32-bit we're fine when reading legacy binlogs as they're fully streaming
   - on 64-bit we just store all strings in memory, and it results in much faster performance
 
The data looks good:
image

### Binlog size reduction:

Strings: | Binlog: | Legacy binlog size (KB) | New binlog size (KB)
-- | -- | -- | --
2,744,490 | vswin.binlog | 703,863 | 274,714
2,711,714 | 790.binlog | 791,488 | 283,364
1,633,734 | 23GB.binlog | 22,558,820 | 172,617
681,346 | mmitche.binlog | 590,938 | 157,915
302,685 | roslyn.binlog | 194,354 | 49,277
288,136 | dotnet.binlog | 64,577 | 24,599
220,177 | vsmac.binlog | 143,178 | 35,095
42,658 | Ide.binlog | 20,753 | 12,334

Playback of binlog into binlog, perf improvement:
Playback of legacy vswin.binlog by legacy: 6:54
Playback of legacy vswin.binlog by new: 2:56
Playback of new vswin.binlog on 32-bit: 3:17
Playback of new vswin.binlog on 64-bit: 2:13

In summary, I expect builds to be faster (up to 4 minutes improvement for really large builds) and binlogs to be smaller (up to 50% for small binlogs, 2-4x for medium size, and 131x (from 23 GB to 172 MB) for our most extreme binlog.

### CODE REVIEW

**Comment by KirillOsenkov** [comment]
The latest viewer version (2.1.300) has already been updated to read the binlog format version 10.

**Comment by KirillOsenkov** [comment]
Hey @mmitche your binlog goes down in size from 590 MB to 157 MB.

@aolszowka I'm probably too late with this, but your binlog goes down from 23 GB to 172 MB! You get the first prize!

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 1]
```diff
@@ -788,5 +1003,89 @@ private void Write(ProfiledLocation e)
             Write(e.ExclusiveTime);
             Write(e.InclusiveTime);
         }
+
+        internal struct HashKey : IEquatable<HashKey>
+        {
+            private ulong value;
```
Comment (drewnoakes) [inline src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:1009]:
```suggestion
        internal readonly struct HashKey : IEquatable<HashKey>
        {
            private readonly ulong value;
```

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 2]
```diff
@@ -788,5 +1003,89 @@ private void Write(ProfiledLocation e)
...
+                {
+                    return Equals(other);
+                }
+
+                return false;
+            }
+
+            public override int GetHashCode()
+            {
+                return (int)value;
```
Comment (drewnoakes) [inline src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:1052]:
```suggestion
                return unchecked((int)value);
```

File: src/Build/Microsoft.Build.csproj [hunk 3]
```diff
@@ -26,6 +26,10 @@
 
     <!-- Do not generate a warning that our 'stable' package should not have a prerelease dependency. -->
     <NoWarn>$(NoWarn);NU5104</NoWarn>
+
+    <!-- Fnv-1a 64-bit hash we use in BuildEventArgsWriter to deduplicate strings uses UInt64 which is not CLS-compliant -->
+    <NoWarn>$(NoWarn);CS3001;CS3002;CS3003</NoWarn>
```
Comment (drewnoakes) [inline src/Build/Microsoft.Build.csproj:31]: Wouldn't it be better to mark the assembly as not CLS compliant? I'm probably missing something here.

**Review by drewnoakes — COMMENTED** [review]
Amazing PR. Just some small things I saw while reading through.

**Comment by aolszowka** [comment]
> @aolszowka I'm probably too late with this, but your binlog goes down from 23 GB to 172 MB! You get the first prize!

This is a HUGE win, I'll make sure to pass it along and keep it in my toolbox :) Thank you!

File: src/Build/Microsoft.Build.csproj (hunk 3 above)
Comment (KirillOsenkov) [reply src/Build/Microsoft.Build.csproj:31]: No, I just don't know about CLS compliance? I'd be happy to. How do I do that?

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs (hunk 1 above)
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:1009]: I was actually thinking of deleting the HashKey struct and just using ulong directly... still on the fence. Thoughts?

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 4]
```diff
@@ -864,28 +1004,68 @@ private IEnumerable ReadItemList()
...
+                return null;
+            }
+            else if (index == 1)
+            {
+                return string.Empty;
+            }
+
+            // we reserve numbers 2-9 for future use.
+            // the writer assigns 10 as the index of the first string
+            index -= BuildEventArgsWriter.StringStartIndex;
```
Comment (rainersigwald) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1060]: Why do this manipulation here rather than just adding the empty slots so the logical and physical indices match?

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 5]
```diff
@@ -972,10 +1152,145 @@ private EvaluationLocation ReadEvaluationLocation()
...
+
+                    if (stream != null)
+                    {
+                        stream.Dispose();
+                        stream = null;
+                    }
+                }
+                catch
+                {
+                }
```
Comment (rainersigwald) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1292]: Comment catch-all please

**Review by rainersigwald — COMMENTED** [review]
I'm **_so_** excited about this! 🙌🏻

Should we add some test cases around deliberately malformed logs (misordered, unused indexes, that sort of thing)?

A few nits but looking great. My only big question is tactical: should we guard this behind a change wave? If we fork now and put this in the very beginning of 16.10 we can probably just make it unconditional, since making the write side conditioned would be a bit inconvenient.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 4 above)
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1060]: Adding the empty slots felt wasteful ;) Also 10 is a number I pulled out of my hat. I doubt we'll use these special slots in the future I'm always scared of painting myself in a corner. Maybe it's just FUD though.

**Comment by KirillOsenkov** [comment]
I'm personally not in a rush for 16.9 so I'd rather just wait until we fork for 16.9 and then just merge into master so we have lots of runway to dogfood for 16.10. This way if we have to tweak the file format we'll have more time.

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 6]
```diff
@@ -718,34 +904,63 @@ private void Write(byte[] bytes)
...
+            binaryWriter.Write(b);
+        }
+
         private void Write(bool boolean)
         {
             binaryWriter.Write(boolean);
         }
 
-        private void Write(string text)
+        private void WriteDeduplicatedString(string text)
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:917]: Technically I could rename Write to WriteRaw (it's only called once) and rename WriteDeduplicatedString back to Write to minimize the diff. But in about half the cases there'll still be a diff because half the cases call WriteOptionalString.

[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:917]: WriteOptionalString is no longer needed because WriteDeduplicatedString has the ability to write null by just using reserved index 0.

[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:917]: The historical reason is I first wasn't sure whether deduping all strings would be worse or better, so I deduped just some of them. But gradually I deduped more and more and it got better and better, so deduping all ended up being optimal. Of course I didn't know that from the outset so I kept a parallel name all the way.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 5 above)
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1292]: If I comment it it won't catch exceptions ;P

Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1292]: j/k will add

File: src/Build/Microsoft.Build.csproj (hunk 3 above)
Comment (KirillOsenkov) [reply src/Build/Microsoft.Build.csproj:31]: pragmas are messy and I'd have to do it in a few places. Any concerns with changing the assembly to CLSCompliant(false)?
https://github.com/dotnet/msbuild/blob/74eb87c20d2495f81788e1bc7bb27203474818f4/src/Build/AssemblyInfo.cs#L39

Unless I'm missing anything CLS compliance was never useful for anything.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 4 above)
Comment (Forgind) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1060]: Are there some strings (other than null and empty) that are far more common than others? I'm wondering if it would reduce the size of the binlog to have a Huffman encoding where the first bit being 0 indicates "special string" and 1 is other, and special means one of the most common strings and involves fewer bits. Probably too complicated to be worth it anyway.

**Review by Forgind — COMMENTED** [review]
Love it!

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 7]
```diff
@@ -49,9 +78,23 @@ public BuildEventArgs Read()
...
+            // just ingest their data and continue.
+            while (IsAuxiliaryRecord(recordKind))
             {
-                ReadBlob(recordKind);
+                // these are ordered by commonality
+                if (recordKind == BinaryLogRecordKind.String)
+                {
+                    ReadStringRecord();
+                }
+                else if (recordKind == BinaryLogRecordKind.NameValueList)
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:90]: Yup, see above: // these are ordered by commonality

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 4 above)
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1060]: Yeah, a bit too complicated. Remember that the entire thing is also compressed by GZipStream.

**Comment by KirillOsenkov** [comment]
I actually discovered that the CLSCompliant warnings have disappeared. Maybe in early iterations the types I added were public and that's why I added the NoWarns? But just removing them now seems to be fine.

File: src/Build/Microsoft.Build.csproj (hunk 3 above)
Comment (KirillOsenkov) [reply src/Build/Microsoft.Build.csproj]: Actually turns out we don't need any of these.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (Forgind) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:90]: 🤐 I promise I know how to read.

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs (hunk 6 above)
[the remark this answers is not in the document]
Comment (rainersigwald) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:920]: Thanks for explaining!

**Review by BenVillalobos — COMMENTED** [review]
Very quick skim, submitting small nits. The results look great 😀

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 4 above)
Comment (ladipro) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1060]: The index is a signed integer so you could have treated all negative numbers as reserved and have the (non-negative) indices for real strings match logical vs. physical.

E.g. `-1` for `null`, `-2` for `String.Empty`

Comment (ladipro) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1060]: Oh, I see, that would waste space because of the 7bit encoding, never mind!

**Review by ladipro — COMMENTED** [review]
Awesome wins and shame on GZipStream that it leaves so much redundant data in it the compressed stream.

Not having encountered hash code collisions in test data doesn't mean that they don't exist. I understand that this is "just logging" and if, with an extremely low probability, the log lies, we are kind of fine with that. Still, isn't it fundamentally a bug? It's unusual for a component to have this probabilistic correctness characteristics in an otherwise precise system. Do you happen to know how much of the wins would be lost if the string lookup was implemented "correctly"? Thank you!

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 8]
```diff
@@ -864,28 +1004,68 @@ private IEnumerable ReadItemList()
...
+            // the writer assigns 10 as the index of the first string
+            index -= BuildEventArgsWriter.StringStartIndex;
+            if (index >= 0 && index < this.stringRecords.Count)
+            {
+                object storedString = stringRecords[index];
+                string result = stringStorage.Get(storedString);
+                return result;
+            }
+
+            return string.Empty;
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1068]: I think you're right, it's better to crash as soon as you noticed that something has gone wrong. I'll add an informative exception.

**Comment by KirillOsenkov** [comment]
@ladipro I totally agree on your point about hash collisions. It's a very valid concern.

Without relying on hashes though, we'd have to hold on to all strings we saw because that's the only way to be absolutely sure the strings match. Holding on to strings is absolutely not an option in a 32-bit process (as we have gigabytes of them), even if we choose a single string per hashcode as a "sample". Writing them to a page file like I do with Reader is an option, but you'd need to constantly seek to read out a string from disk for every new string you encounter. It would absolutely kill perf and defeat the purpose.

I'm afraid that if we want the perf wins, we have to allow a small possibility of a collision, where in a random binlog (1) a random string you happened to be looking at (2) is switched to another random string. You only look at a tiny small fraction of all strings in the binlog (there are millions), and chances are that you happened to look at the corrupt one are infinitesimally small (given that I haven't seen any collisions in our top 10 largest binlogs). I say it's worth the risk!

**Comment by KirillOsenkov** [comment]
I guess we could also store the string length and the first character (in addition to hash), to further drastically reduce the chance of collision. But honestly I don't think it's even worth it.

**Comment by cdmihai** [comment]
Regarding the hash problem and the large memory required to store the strings, wouldn't using msbuild's string interner (maybe using @ladipro's [rewritten version of it](https://github.com/dotnet/msbuild/pull/5663)) fix the problem? The interner also dedupes strings, so the in-memory string size should be the same as the deduped string size from the binlog

**Comment by aolszowka** [comment]
> Writing them to a page file like I do with Reader is an option, but you'd need to constantly seek to read out a string from disk for every new string you encounter. It would absolutely kill perf and defeat the purpose.

If there was a big concern about a collision what about a post process step? So you opt in to "do it fast with a hash" but allow a slower pass that captured all the strings then performed this process, at that point if there was a collision you could bark, but if there was no collision you reap the gains when sharing the binlog?

Does nothing for the perf though which is what I assume was driving a lot of this?

I suspect the big concern is trying to capture this one off, realistically though by the time we're digging though the bin logs its probably reproducible on our end so its not like we're totally out of luck.

**Comment by ladipro** [comment]
> Regarding the hash problem and the large memory required to store the strings, wouldn't using msbuild's string interner (maybe using @ladipro's rewritten version of it) fix the problem? 

The interner holds on to strings only as long as something else is keeping them alive so it wouldn't guarantee correctness either. It's basically a cache of recently seen and not yet collected strings, not of all strings.

**Comment by ladipro** [comment]
> I guess we could also store the string length and the first character (in addition to hash), to further drastically reduce the chance of collision. But honestly I don't think it's even worth it.

Agreed that it wouldn't be worth it. It's basically like using a wider hashcode but with poorer collision characteristics.

> I'm afraid that if we want the perf wins, we have to allow a small possibility of a collision, where in a random binlog (1) a random string you happened to be looking at (2) is switched to another random string. You only look at a tiny small fraction of all strings in the binlog (there are millions), and chances are that you happened to look at the corrupt one are infinitesimally small (given that I haven't seen any collisions in our top 10 largest binlogs). I say it's worth the risk!

Let's do it!

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 9]
```diff
@@ -972,10 +1152,149 @@ private EvaluationLocation ReadEvaluationLocation()
...
+                    filePath = Path.GetTempFileName();
+                    var utf8noBom = new UTF8Encoding(encoderShouldEmitUTF8Identifier: false);
+                    stream = new FileStream(
+                        filePath,
+                        FileMode.OpenOrCreate,
+                        FileAccess.ReadWrite,
+                        FileShare.None,
+                        bufferSize: 4096,
+                        FileOptions.RandomAccess | FileOptions.DeleteOnClose);
+                    streamWriter = new StreamWriter(stream, utf8noBom, 65536);
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1208]: Done

File: src/Build/Logging/BinaryLogger/BinaryLogger.cs [hunk 10]
```diff
@@ -137,6 +143,7 @@ public void Initialize(IEventSource eventSource)
             }
 
             stream = new GZipStream(stream, CompressionLevel.Optimal);
+            stream = new BufferedStream(stream, bufferSize: 32768);
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BinaryLogger.cs:150]: Done

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs]: I changed it to throw an exception.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 11]
```diff
@@ -972,10 +1161,152 @@ private EvaluationLocation ReadEvaluationLocation()
...
+
+                var position = (StringPosition)storedString;
+
+                stream.Position = position.FilePosition;
+                stringBuilder.Length = position.StringLength;
+                for (int i = 0; i < position.StringLength; i++)
+                {
+                    char ch = (char)streamReader.Read();
+                    stringBuilder[i] = ch;
+                }
```
Comment (ladipro) [inline src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1276]: nit: Why not use `StreamReader.Read(Char[], Int32, Int32)` to read into an array and then `new String(Char[])`?

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 12]
```diff
@@ -635,26 +744,29 @@ private void WriteProperties(IEnumerable properties)
                 return;
             }
 
+            if (nameValueListBuffer.Count > 0)
```
Comment (ladipro) [inline src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:747]: super-nit: Just `Clear()` unconditionally for brevity?

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs [hunk 13]
```diff
@@ -718,34 +907,63 @@ private void Write(byte[] bytes)
...
+
+            return (recordId, hash);
+        }
+
+        private void WriteStringRecord(string text)
+        {
+            // Switch the binaryWriter used by the Write* methods to the direct underlying stream writer.
+            // We want this record to precede the record we're currently writing to currentRecordWriter
+            // which is backed by a MemoryStream buffer
+            binaryWriter = this.originalBinaryWriter;
```
Comment (ladipro) [inline src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:960]: Worth adding something like this, given that the pattern is repeated throughout this file?

```C#
        private struct RedirectionHolder : IDisposable
        {
            private readonly BuildEventArgsWriter _writer;

            public RedirectionHolder(BuildEventArgsWriter writer)
            {
                _writer = writer;
                _writer.binaryWriter = _writer.originalBinaryWriter;
            }

            public void Dispose()
            {
                _writer.binaryWriter = _writer.currentRecordWriter;
            }
        }
```

It would then allow you to write:

```suggestion
            using var redirection = RedirectWritesToOriginalBinaryWriter();
```

**Review by ladipro — COMMENTED** [review]
Left a few comments, mostly nits.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 14]
```diff
@@ -972,10 +1161,152 @@ private EvaluationLocation ReadEvaluationLocation()
...
+                // seek to retrieve them. Instead we'll keep storing small strings
+                // in memory until we allocate 2 GB. After that, all strings go to
+                // the file.
+                // Win-win: small binlog playback is fast and large binlog playback
+                // doesn't OOM.
+                if (text.Length <= StringSizeThreshold && totalAllocatedShortStrings < 2_000_000_000)
+                {
+                    // note that we write strings in UTF8 so we don't need to multiply by 2 as chars
+                    // will be 1 byte on average
+                    totalAllocatedShortStrings += text.Length;
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1247]: MSBuild sets the LargeAddressAware flag that lets it address up to 4GB. I've seen MSBuild.exe easily with 2.8 GB commit memory. The thinking here is that MSBuild.exe can't be doing anything other than binlog playback so we're free to spend most of the process memory on caching as many strings we can.
However your point that strings are two bytes per char in memory is valid, so I reduced to 1 billion instead.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 11 above)
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1276]: Then I'd have to manage the char array myself (reuse to avoid reallocating and preallocated proper size). This way I let StringBuilder manage and reuse the char array chains it allocates. Once our StringBuilder reaches the threshold length here, it continues with no allocations except to create the final string.

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs (hunk 12 above)
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs:747]: Clear() also increments the internal `version` field of the list. I'm sure it doesn't matter at all and it's OK either way, but I thought of the version field overflowing (it's an int32) and decided to do this instead :) Call me paranoid :)

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs (hunk 13 above)
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs]: Neat idea, done.

File: src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs (hunk 12 above)
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsWriter.cs]: Haah, you're right. Simpler is better. Fixed.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 15]
```diff
@@ -972,10 +1161,152 @@ private EvaluationLocation ReadEvaluationLocation()
...
+            public void Dispose()
+            {
+                try
+                {
+                    if (streamWriter != null)
+                    {
+                        streamWriter.Dispose();
+                        streamWriter = null;
+                    }
+
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs:1293]: Nothing interesting happens in StreamReader.Dispose(): https://referencesource.microsoft.com/#mscorlib/system/io/streamreader.cs,011147a163b10458

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs [hunk 16]
```diff
@@ -972,10 +1161,152 @@ private EvaluationLocation ReadEvaluationLocation()
...
+        /// Stores large strings in a temp file on disk, to avoid keeping all strings in memory.
+        /// Only creates a file for 32-bit MSBuild.exe, just returns the string directly on 64-bit.
+        /// </summary>
+        internal class StringStorage : IDisposable
+        {
+            private string filePath;
+            private FileStream stream;
+            private StreamWriter streamWriter;
+            private StreamReader streamReader;
+            private StringBuilder stringBuilder;
```
[the remark this answers is not in the document]
Comment (KirillOsenkov) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs]: Fixed

**Comment by KirillOsenkov** [comment]
I think I've answered or resolved all comments, but let me know if you have more or disagree with some of my resolutions. Meanwhile I'll do some more final testing on it.

File: src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs (hunk 14 above)
[the remark this answers is not in the document]
Comment (aolszowka) [reply src/Build/Logging/BinaryLogger/BuildEventArgsReader.cs]: @KirillOsenkov 

>MSBuild sets the LargeAddressAware flag that lets it address up to 4GB. I've seen MSBuild.exe easily with 2.8 GB commit memory. The thinking here is that MSBuild.exe can't be doing anything other than binlog playback so we're free to spend most of the process memory on caching as many strings we can.

I would be careful with this thought. So sometimes there is a poorly written MSBuild Project extension that does all types of nasty things (such as a poor logger or something else), or if they cache a large number of properties or projects. We were encountering the 3.8gb or whatever it was limitation in 32bit MSBuild causing us to switch to MSBuild64 to allow us to address more memory. It was not unusual to see msbuild.exe usage in the 8-9gb territory for us per node (multiply that out by the number of nodes when using `/maxCPU`). 

Basically the third party vendor was doing the equivalent of this, for every single project the changing the properties polluting the cache:

`ProjectCollection.LoadProject(string)`

https://docs.microsoft.com/en-us/dotnet/api/microsoft.build.evaluation.projectcollection.loadproject?view=msbuild-16-netcore#Microsoft_Build_Evaluation_ProjectCollection_LoadProject_System_String_

While I believe the vendor eventually fixed it it took them well over 4 years to do so, and I still think they have not fixed everywhere they did this pattern. @madkat.

**Review by ladipro — APPROVED** [review]
Thank you!

**Comment by KirillOsenkov** [comment]
I'm happy to report a nice performance improvement from this PR.
master is 32s with /bl, this PR is 26s.
Pure MSBuild without any logging is 15s, with text file diagnostic log 55s.

master | #6017 | no /bl | /flp1:v=diag
-- | -- | -- | --
32.377 | 27.737 | 15.476 | 55.61
31.758 | 26.561 | 15.416 | 56.145
31.481 | 26.486 | 15.536 | 55.72

**Comment by KirillOsenkov** [comment]
Shall we merge this?

**Comment by Forgind** [comment]
Yep!

I'm still trying to merge things in a gradual way, but I'll get to your other PR in the next few days.
